src/gui/transferlistwidget.cpp

```diff
@@ -59,7 +61,124 @@
 #include "transferlistsortmodel.h"
 #include "updownratiodlg.h"
 
-static QStringList extractHashes(const QList<BitTorrent::TorrentHandle *> &torrents);
+namespace
+{
+    QStringList extractHashes(const QList<BitTorrent::TorrentHandle *> &torrents)
+    {
+        QStringList hashes;
+        foreach (BitTorrent::TorrentHandle *const torrent, torrents)
+            hashes << torrent->hash();
+
+        return hashes;
+    }
+
+    // A checkable menu item that supports tristate checkboxes.
+    class CheckBoxMenuItem: public QWidgetAction
+    {
+    public:
+        using ToggleFn = std::function<void (Qt::CheckState)>;
+
+        CheckBoxMenuItem(const QString &text, const ToggleFn &onToggle, Qt::CheckState initialState, QObject *parent);
+
+    private:
+        static constexpr int hMargin = 3;
+        static constexpr int vMargin = 1;
+
+        class IconHelper: public QCheckBox
```

Comment:
Done. Although I see the suggestion has since gotten a thumbs down, so please work out any disagreement. I don't have strong feelings either way.